src/zjs_fs.c

```diff
@@ -0,0 +1,1048 @@
+        return jerry_create_boolean(true);
+    } else {
+        return jerry_create_boolean(false);
+    }
+}
+
+static jerry_value_t is_directory(const jerry_value_t function_obj,
+                                  const jerry_value_t this,
+                                  const jerry_value_t argv[],
+                                  const jerry_length_t argc)
+{
+    struct fs_dirent* entry;
+
+    if (!jerry_get_object_native_handle(this, (uintptr_t*)&entry)) {
+        return zjs_error("native handle not found");
+    }
+    if (entry->type == FS_DIR_ENTRY_DIR) {
+        return jerry_create_boolean(true);
+    } else {
+        return jerry_create_boolean(false);
+    }
+}
+
+static void free_stats(const uintptr_t native)
+{
+    struct zfs_dirent* entry = (struct zfs_dirent*)native;
+    if (entry) {
+        zjs_free(entry);
+    }
+}
+
+static jerry_value_t create_stats_obj(struct fs_dirent* entry)
+{
+
+    jerry_value_t stats_obj = jerry_create_object();
+
+    struct fs_dirent* new_entry = zjs_malloc(sizeof(struct fs_dirent));
+    if (!new_entry) {
+        return zjs_error("malloc failed");
+    }
+    memcpy(new_entry, entry, sizeof(struct fs_dirent));
+
+    jerry_set_object_native_handle(stats_obj, (uintptr_t)new_entry, free_stats);
+
+    zjs_obj_add_function(stats_obj, is_file, "isFile");
+    zjs_obj_add_function(stats_obj, is_directory, "isDirectory");
+
+    return stats_obj;
+}
+
+static jerry_value_t zjs_open(const jerry_value_t function_obj,
+                              const jerry_value_t this,
+                              const jerry_value_t argv[],
+                              const jerry_length_t argc,
+                              uint8_t async)
+{
+    if (argc < 3 && async) {
+        return invalid_args();
+    }
+    if (!jerry_value_is_string(argv[0])) {
+        return invalid_args();
+    }
+    if (!jerry_value_is_string(argv[1])) {
+        return invalid_args();
+    }
+    if (async) {
+        if (!jerry_value_is_function(argv[argc - 1])) {
+            return invalid_args();
+        }
+    }
+
+    file_handle_t* handle = new_file();
+    handle->fd_val = jerry_create_object();
+
+    jerry_size_t size = MAX_PATH_LENGTH;
+    char path[size];
+
+    zjs_copy_jstring(argv[0], path, &size);
+    if (!size) {
+        return zjs_error("size mismatch");
+    }
+
+    size = 4;
+    char mode[size];
+
+    zjs_copy_jstring(argv[1], mode, &size);
+    if (!size) {
+        return zjs_error("size mismatch");
+    }
+
+    DBG_PRINT("Opening file: %s, mode: %s\n", path, mode);
+
+    handle->mode = get_mode(mode);
+
+    /*
+     * TODO: Currently Zephyr has no concept of file permissions, when you open
+     *       a file it is automatically readable and writable. This comment
+     *       is left as a reminder that if other systems/OS have permissions
+     *       we may want to implement the logic, but for now the 'mode' parameter
+     *       is unused.
+     */
```

Comment:
You're right that there are no permissions. If there were, then opening for read or write might fail because of permissions errors. But the modes are not really about permissions, they are about what the user intends to do with the file. So we can still support them. For example, if the user asks to open a file for reading and then tries to write to it, this should fail. The modes are also important in terms of where writes will occur. If they use 'a', then writes should always go to the end of the file and no allow them to seek backwards.

Here's the easy way out... to land your patch quickly, we need the bare minimum of support for 'r' and 'w'. With those, the user can do anything, though certain operations will be a big pain because 'w' wipes out (truncates) the file if it exists. So get those two working, fail open if anything else is passed, and document that those are the only things supported at present. Add test cases that prove that they're working correctly:

For 'w' mode:
* Prove open will create a new file when it doesn't exist
* Prove open will overwrite a file that does exist
* Prove the file contents can be read back with an 'r', so they were written correctly.
* Prove that read() fails on this descriptor.
* Prove that you can seek around and modify different parts (somewhat pointless in this mode, but valid).

For 'r' mode:
* Prove that open fails with 'an exception' if the file doesn't exist.
* Prove that you can seek around and read different parts.
* Prove that write() fails.

Worth testing: can you open the same file for writing and then for reading with a different descriptor, open at the same time? Can you then make changes with the write descriptor and immediately observe them with the read one?